### ISSUE TITLE
`on_shutdown` callback not called on Windows 10

### ISSUE THREAD

**Issue by chhinze** [zauberzeug/nicegui#1050]
### Description

Hi all, 

thanks for this awesome project.

Registering a callback with `app.on_shutdown` should lead to the callback to be called during hot reloading (described in  <https://nicegui.io/documentation#events>). This worked on a machine with Ubuntu 23.04 as expected, but on Windows 10 it is not called.

Minimal example to reproduce:

```py
from nicegui import app, ui

app.on_shutdown(lambda: print("Shutdown called"))
app.on_startup(lambda : print("Startup called"))

ui.markdown("# Title")
ui.run(title="Does not call shutdown on Win10 Demo")
```

On Windows 10 the console Output is

```
Startup called
NiceGUI ready to go in http://...
?[33mWARNING?[0m: Watch files detected changes in ...

[here is, where the reload happens, triggered by Ctrl+S in the file.]

Startup called
NiceGUI ready to go in http://...
```

In Ubuntu I see a line `Shutdown called` before the second `startup called`, this is also what I would expect reading the documentation. 

Am I using it wrong?
Is this related to #604 i.e., the Ctrl handler is not set properly?

Best regards 
Chris

**Comment by falkoschindler** [comment]
Thanks for reporting this issue, @chhinze! This is definitely unexpected.

NiceGUI's lifecycle hooks like `on_startup` and `on_shutdown` rely on FastAPI. Can you, please, check if the following pure FastAPI example has the same problem on Windows 10?

main.py:
```py
import uvicorn
from fastapi import FastAPI

app = FastAPI()

@app.on_event("startup")
async def startup_event():
    print("Application startup")

@app.on_event("shutdown")
async def shutdown_event():
    print("Application shutdown")

@app.get("/")
def read_root():
    return {"Hello": "World"}

if __name__ == "__main__":
    uvicorn.run('main:app', host="0.0.0.0", port=8000, reload=True)
```

**Comment by chhinze** [comment]
Thank you a lot for the quick response, @falkoschindler.

I just tested your code. The problem persists also with the pure FastAPI example you provided. The output is:

```
INFO:     Will atch for changes in these directories: ['...']
INFO:     Uvicorn running on http://0.0.0.0:8000 (Press CTRL+C to quit)
INFO:     Started reloader process [3644] using WatchFiles
INFO:     Started server process [7404]
INFO:     Waiting for application startup.
Application startup
INFO:     Application startup complete.
WARNING:  WatchFiles detected changes in 'main.py'. Reloading...
INFO:     Started server process [5596]
INFO:     Waiting for application startup.
Application startup
INFO:     Application startup complete.
```

On Linux, after the line `WARNING:  WatchFiles detected changes in 'main.py'. Reloading...` I get the additional lines 

```
INFO:   Shutting down
INFO:   Waiting for application shutdown.
Application shutdown 
INFO:   Application shutdown complete
INFO:   Finished server process [ProcessNo]
```

Thanks to your information that it is related to FastAPI I now found the open issues [fastapi#6026](https://github.com/tiangolo/fastapi/discussions/6026)  and [fastapi#6391](https://github.com/tiangolo/fastapi/discussions/6391). A fix seems to be [merged already](https://github.com/encode/uvicorn/pull/1584)  back in April and is scheduled for release 0.22.0. 

I guess the best option is to just wait for this release of  FastAPI. 
Or is there a workaround to register a callback before reloading?

**Comment by rodja** [comment]
> Or is there a workaround to register a callback before reloading?

Unfortunately not. Or none that we are aware of. We would really much like to execute some code before uvicorn reloads, but we have not found a way yet.

**Comment by falkoschindler** [comment]
Ok, we'll wait for https://github.com/encode/uvicorn/pull/1584 to be released.

**Comment by chhinze** [comment]
Yes, you are right. I was misguided by the milestone progress bar in the issue ([encode/uvicorn#1584](https://github.com/encode/uvicorn/pull/1584)). Sorry for the confusion.
I also tested manually upgrading uvicorn: `on_shutdown` callbacks work as expected now.

**Comment by firai** [comment]
@falkoschindler Can uvicorn be upgraded to 0.22.0? Or would there be blocking conflicts from changes in uvicorn 0.21.0 or 0.22.0?

**Comment by firai** [comment]
Ah, I assumed that development was on `main` without looking. Didn't realize that it's already changed on the 1.3 branch. Thanks!

**Comment by rodja** [comment]
Yea, we normally do everything on `main`. But breaking changes have accumulated on the v1.3 branch and is almost ready to be released.

**Comment by falkoschindler** [comment]
Ok, uvicorn has been updated in commit https://github.com/zauberzeug/nicegui/commit/578de6a09101b0e946eba71a879e81a195de416b. It will be released with NiceGUI 1.3.0. Therefore I'll close this issue.
